File: sdks/ruby/lib/kafka_do/kafka.rb
# frozen_string_literal: true

module KafkaDo
  # Main Kafka client class
  #
  # Provides a ruby-kafka compatible API for interacting with Kafka brokers
  # through the kafka.do RPC service.
  #
  # @example Creating a Kafka client
  #   kafka = Kafka.new(['broker1:9092', 'broker2:9092'])
  #
  # @example With options
  #   kafka = Kafka.new(['broker1:9092'],
  #     client_id: 'my-app',
  #     connection_url: 'https://kafka.do'
  #   )
  #
  class Kafka
    # Compression types
    COMPRESSION_NONE = :none
    COMPRESSION_GZIP = :gzip
    COMPRESSION_SNAPPY = :snappy
    COMPRESSION_LZ4 = :lz4
    COMPRESSION_ZSTD = :zstd

    # Offset reset policies
    OFFSET_EARLIEST = :earliest
    OFFSET_LATEST = :latest

    # Acknowledgment levels
    ACKS_NONE = 0
    ACKS_LEADER = 1
    ACKS_ALL = -1

    # @return [Array<String>] the configured broker addresses
    attr_reader :brokers

    # @return [String, nil] the client ID
    attr_reader :client_id

    # @return [String] the connection URL for the kafka.do RPC service
    attr_reader :connection_url

    # Create a new Kafka client
    #
    # @param brokers [Array<String>] list of broker addresses (host:port)
    # @param client_id [String, nil] optional client identifier
    # @param connection_url [String] URL for the kafka.do service
    # @param timeout [Integer] connection timeout in seconds
    # @param api_key [String, nil] optional API key for authentication
    def initialize(brokers = [], client_id: nil, connection_url: 'https://kafka.do',
                   timeout: 30, api_key: nil)
      @brokers = Array(brokers)
      @client_id = client_id
      @connection_url = connection_url
      @timeout = timeout
      @api_key = api_key
      @rpc_client = nil
      @closed = false
      @mutex = Mutex.new
    end

    # Set the RPC client (primarily used for testing)
    #
    # @param client [Object] the RPC client instance
    def rpc_client=(client)
      @mutex.synchronize { @rpc_client = client }
    end

    # Get the RPC client, connecting if necessary
    #
    # @return [Object] the RPC client
    # @raise [ConnectionError] if connection fails
    def rpc_client
      @mutex.synchronize do
        raise ClosedError, 'Client is closed' if @closed

        @rpc_client ||= connect_rpc
      end
    end

    # Create a new producer
    #
    # @param compression_codec [Symbol] compression type (:none, :gzip, :snappy, :lz4, :zstd)
    # @param required_acks [Integer] number of acks required (0, 1, or -1 for all)
    # @param max_buffer_size [Integer] maximum buffer size in bytes
    # @param max_buffer_bytesize [Integer] maximum buffer byte size
    # @return [Producer] a new producer instance
    #
    # @example Creating a producer
    #   producer = kafka.producer
    #   producer.produce('Hello!', topic: 'my-topic')
    #   producer.deliver_messages
    #
    def producer(compression_codec: COMPRESSION_NONE,
                 required_acks: ACKS_LEADER,
                 max_buffer_size: 1000,
                 max_buffer_bytesize: 10_000_000,
                 **options)
      raise ClosedError, 'Client is closed' if @closed

      Producer.new(
        self,
        compression_codec: compression_codec,
        required_acks: required_acks,
        max_buffer_size: max_buffer_size,
        max_buffer_bytesize: max_buffer_bytesize,
        **options
      )
    end

    # Create a new consumer
    #
    # @param group_id [String] the consumer group ID
    # @param offset_commit_interval [Integer] offset commit interval in seconds
    # @param offset_commit_threshold [Integer] offset commit threshold
    # @param heartbeat_interval [Integer] heartbeat interval in seconds
    # @param session_timeout [Integer] session timeout in seconds
    # @return [Consumer] a new consumer instance
    #
    # @example Creating a consumer
    #   consumer = kafka.consumer(group_id: 'my-group')
    #   consumer.subscribe('my-topic')
    #   consumer.each_message { |msg| puts msg.value }
    #
    def consumer(group_id:,
                 offset_commit_interval: 10,
                 offset_commit_threshold: 0,
                 heartbeat_interval: 10,
                 session_timeout: 30,
                 **options)
      raise ClosedError, 'Client is closed' if @closed

      Consumer.new(
        self,
        group_id: group_id,
        offset_commit_interval: offset_commit_interval,
        offset_commit_threshold: offset_commit_threshold,
        heartbeat_interval: heartbeat_interval,
        session_timeout: session_timeout,
        **options
      )
    end

    # Create a new admin client
    #
    # @return [Admin] a new admin client instance
    def admin
      raise ClosedError, 'Client is closed' if @closed

      Admin.new(self)
    end

    # Create a new stream processor for a topic
    #
    # Provides a Kafka Streams-like DSL for building stream processing pipelines.
    #
    # @param topic [String] the source topic
    # @param group_id [String] optional consumer group ID
    # @return [Stream] a new stream processor
    #
    # @example Basic stream processing
    #   kafka.stream('orders')
    #     .filter { |msg| msg.value['status'] == 'completed' }
    #     .map { |msg| msg.value['total'] }
    #     .to('order-totals')
    #
    def stream(topic, group_id: nil)
      raise ClosedError, 'Client is closed' if @closed

      Stream.new(self, topic, group_id: group_id)
    end

    # Create a topic (convenience method delegating to Admin)
    #
    # @param topic [String] the topic name
    # @param num_partitions [Integer] number of partitions
    # @param replication_factor [Integer] replication factor
    # @param config [Hash] additional topic configuration
    # @return [void]
    #
    # @example Creating a topic
    #   kafka.create_topic('new-topic', num_partitions: 3, replication_factor: 2)
    #
    def create_topic(topic, num_partitions: 1, replication_factor: 1, config: {})
      admin.create_topic(topic,
                         num_partitions: num_partitions,
                         replication_factor: replication_factor,
                         config: config)
    end

    # Delete a topic (convenience method delegating to Admin)
    #
    # @param topic [String] the topic name
    # @return [void]
    def delete_topic(topic)
      admin.delete_topic(topic)
    end

    # List topics (convenience method delegating to Admin)
    #
    # @return [Array<String>] list of topic names
    def topics
      admin.list_topics
    end

    # Get partitions for a topic
    #
    # @param topic [String] the topic name
    # @return [Array<Integer>] list of partition numbers
    def partitions_for(topic)
      admin.partitions_for(topic)
    end

    # Close the Kafka client
    #
    # @return [void]
    def close
      @mutex.synchronize do
        return if @closed

        @rpc_client&.close rescue nil
        @rpc_client = nil
        @closed = true
      end
    end

    # Check if the client is closed
    #
    # @return [Boolean]
    def closed?
      @closed
    end

    private

    def connect_rpc
      # In production, this would connect via the RPC transport
      # For now, raise an error if no client is set
      raise ConnectionError, 'RPC client not configured. Set via kafka.rpc_client='
    end
  end
end

File: sdks/python/src/kafka_do/kafka.py
"""
Main Kafka class for kafka-do SDK.

Provides the entry point for creating producers, consumers, and admin clients.
"""

from __future__ import annotations

from typing import Any

from .admin import Admin
from .consumer import Consumer
from .producer import Producer
from .types import (
    CompressionType,
    IsolationLevel,
    KafkaConnectionError,
    OffsetResetPolicy,
)


class Kafka:
    """
    Main Kafka client class.

    The Kafka class serves as the entry point for creating producers,
    consumers, and admin clients that connect to Kafka via rpc-do.

    Example:
        # Create Kafka client
        kafka = Kafka(brokers=['localhost:9092'])

        # Create producer
        producer = kafka.producer()
        await producer.connect()
        await producer.send('my-topic', value=b'Hello!')
        await producer.disconnect()

        # Create consumer
        consumer = kafka.consumer(group_id='my-group')
        await consumer.connect()
        await consumer.subscribe(['my-topic'])
        async for message in consumer:
            print(message.value)
        await consumer.disconnect()

        # Create admin
        admin = kafka.admin()
        await admin.connect()
        topics = await admin.list_topics()
        await admin.disconnect()
    """

    __slots__ = (
        "_brokers",
        "_client_id",
        "_timeout",
        "_client",
        "_connection_url",
    )

    def __init__(
        self,
        brokers: list[str] | None = None,
        *,
        client_id: str | None = None,
        timeout: float = 30.0,
        connection_url: str | None = None,
    ) -> None:
        """
        Initialize the Kafka client.

        Args:
            brokers: List of broker addresses (e.g., ['localhost:9092'])
            client_id: Optional client identifier
            timeout: Connection timeout in seconds
            connection_url: Optional direct connection URL for rpc-do
                            (defaults to kafka.do)
        """
        self._brokers = brokers or ["localhost:9092"]
        self._client_id = client_id
        self._timeout = timeout
        self._client: Any = None
        self._connection_url = connection_url or "kafka.do"

    @property
    def brokers(self) -> list[str]:
        """Return the configured broker addresses."""
        return self._brokers.copy()

    @property
    def client_id(self) -> str | None:
        """Return the client ID."""
        return self._client_id

    async def _get_client(self) -> Any:
        """
        Get or create the rpc-do client.

        Returns:
            Connected rpc-do client

        Raises:
            KafkaConnectionError: If connection fails
        """
        if self._client is not None:
            return self._client

        try:
            from rpc_do import connect  # pragma: no cover

            self._client = await connect(  # pragma: no cover
                self._connection_url,
                timeout=self._timeout,
            )

            # Configure the Kafka service with brokers
            await self._client.kafka.configure(  # pragma: no cover
                brokers=self._brokers,
                clientId=self._client_id,
            )

            return self._client  # pragma: no cover
        except ImportError as e:
            raise KafkaConnectionError(
                "rpc-do package not installed. Install with: pip install rpc-do"
            ) from e
        except Exception as e:  # pragma: no cover
            raise KafkaConnectionError(f"Failed to connect to Kafka service: {e}") from e  # pragma: no cover

    async def close(self) -> None:
        """
        Close the Kafka client and release resources.
        """
        if self._client is not None:
            try:
                await self._client.close()
            except Exception:
                pass
            self._client = None

    def producer(
        self,
        *,
        acks: int | str = 1,
        compression_type: CompressionType = CompressionType.NONE,
        batch_size: int = 16384,
        linger_ms: int = 0,
        client_id: str | None = None,
    ) -> Producer:
        """
        Create a new producer.

        Args:
            acks: Number of acknowledgments required (0, 1, or 'all')
            compression_type: Compression type for messages
            batch_size: Maximum batch size in bytes
            linger_ms: Time to wait before sending batched messages
            client_id: Optional client identifier (overrides Kafka client_id)

        Returns:
            New Producer instance
        """
        return Producer(
            self,
            acks=acks,
            compression_type=compression_type,
            batch_size=batch_size,
            linger_ms=linger_ms,
            client_id=client_id or self._client_id,
        )

    def consumer(
        self,
        group_id: str | None = None,
        *,
        auto_commit: bool = True,
        auto_commit_interval_ms: int = 5000,
        session_timeout_ms: int = 10000,
        heartbeat_interval_ms: int = 3000,
        max_poll_records: int = 500,
        max_poll_interval_ms: int = 300000,
        isolation_level: IsolationLevel = IsolationLevel.READ_UNCOMMITTED,
        auto_offset_reset: OffsetResetPolicy = OffsetResetPolicy.LATEST,
        client_id: str | None = None,
    ) -> Consumer:
        """
        Create a new consumer.

        Args:
            group_id: Consumer group ID (required for group coordination)
            auto_commit: Enable auto-commit of offsets
            auto_commit_interval_ms: Interval for auto-commit in milliseconds
            session_timeout_ms: Session timeout for group membership
            heartbeat_interval_ms: Heartbeat interval for group coordination
            max_poll_records: Maximum records per poll
            max_poll_interval_ms: Maximum interval between polls
            isolation_level: Transaction isolation level
            auto_offset_reset: Offset reset policy for new consumers
            client_id: Optional client identifier (overrides Kafka client_id)

        Returns:
            New Consumer instance
        """
        return Consumer(
            self,
            group_id=group_id,
            auto_commit=auto_commit,
            auto_commit_interval_ms=auto_commit_interval_ms,
            session_timeout_ms=session_timeout_ms,
            heartbeat_interval_ms=heartbeat_interval_ms,
            max_poll_records=max_poll_records,
            max_poll_interval_ms=max_poll_interval_ms,
            isolation_level=isolation_level,
            auto_offset_reset=auto_offset_reset,
            client_id=client_id or self._client_id,
        )

    def admin(
        self,
        *,
        request_timeout_ms: int = 30000,
    ) -> Admin:
        """
        Create a new admin client.

        Args:
            request_timeout_ms: Request timeout in milliseconds

        Returns:
            New Admin instance
        """
        return Admin(
            self,
            request_timeout_ms=request_timeout_ms,
        )

    async def __aenter__(self) -> Kafka:
        """Async context manager entry."""
        await self._get_client()
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: Any,
    ) -> None:
        """Async context manager exit."""
        await self.close()
